yr942-xl3725/project.md

```diff
@@ -0,0 +1,53 @@
+---
+title: "Build A Job Scheduler for Clusters based on Reinforcement Learning"
+author:
+  - Yafu Ruan <yr942@nyu.edu>
+  - Xuan Li <xl3725@nyu.edu>
+---
+
+> In what follows instructions (including this one) are in blockquotes, while
+> text not in blockquotes is a sample. You should of course replace the title
+> author metadata that shows up above with appropriate values.
+>
+> You should start your proposal by stating a problem statement or an overall
+> goal  as below.
+
+For our Big Data and Machine Learning final project we plan to develop a customized scheduler policy using Reinforcement Learning based on an existing paper, <cite>[RLSK: A Job Scheduler for Federated Kubernetes Clusters based on Reinforcement Learning][1]</cite>.
+
+[1]: https://conferences.computer.org/cpsiot/pdfs/IC2E2020-6j9eixwcPIo0tUoMnoCahH/109900a116/109900a116.pdf
+
+# Motivation
+> Next you should talk about why, citing sources for your beliefs.
+
+The default algorithm of kube-scheduler can help us in most cases, but sometimes people need to develop their own scheduling policy and change the configurations for specific projects in different circumstances. This process involves lots of repeated work done manually to get the optimized solution. To improve the efficiency of the whole process, machine learning models are considered to assist making decisions when scheduling. The status of the clusters and the requirement of the job could be some state features as input, and the model will decide how to allocate the resources to those jobs. A metric will be designed to analyze whether the decision is actually a “good choice”. Reinforcement Learning is one of the models that could explain the problem well.
+
+# Proposed Solution
+> Next you should talk about how you plan to address the problem. 
+> This should include information about what infrastructure (e.g., CloudLab
+> resources) you plan to use, how you plan to make progress, and how you plan to
+> evaluate your view.
+
+We plan to construct a Reinforcement Learning model based information from Kubernetes API and job requirements.
+
+* Firstly we design the basic training algorithm using pytorch.
+* Next, we will collect resource information from Kubernetes API and transform them to vectors in a proper way. We combine it with job information, which will also be a vector, and take it as the state variable.
+* After that, we will design the reward metric to determine how “good” the decision is made by the model. The metric needs to consider the trade-off between the overall utilization and the balance of resources and clusters.
+
+We plan to run our experiments on multiple CloudLab nodes.
```

Comment:
What workloads do you plan to use to evaluate your scheduler?

Reply:
In the testing process, we are going to simulate job loads, whose arrival time, duration and recourses come from certain distributions. This should be similar to the work load simulated in DeepRM as described below(https://people.csail.mit.edu/alizadeh/papers/deeprm-hotnets16.pdf):

“jobs arrive online according to a Bernoulli process. The average job arrival rate is chosen such that the average load varies between 10% to 190% of cluster capacity. We assume two resources, i.e., with capacity {1r, 1r}. Job durations and resource demands are chosen as follows: 80% of the jobs have duration uniformly chosen between 1t and 3t; the remaining are chosen uniformly from 10t to 15t. Each job has a dominant resource which is picked independently at random. The demand for the dominant resource is chosen uniformly between 0.25r and 0.5r and the demand of the other resource is chosen uniformly between 0.05r and 0.1r. “

We will try to figure whether this is feasible.

Reply:
Can you add the information in the comment above to the proposal itself.